File: ejemplos-chat.js
/**
 * Ejemplos de código JavaScript para practicar con GitHub Copilot Chat
 * Usa estos ejemplos para hacer preguntas al chat y mejorar tu código
 */

// Ejemplo 1: Función que necesita modernización
function sumarArray(arr) {
    var total = 0;
    for (var i = 0; i < arr.length; i++) {
        total += arr[i];
    }
    return total;
}
// Pregunta al chat: "Moderniza esta función usando ES6+"


// Ejemplo 2: Callback hell que necesita promesas
function obtenerDatos(userId, callback) {
    obtenerUsuario(userId, function(usuario) {
        obtenerPosts(usuario.id, function(posts) {
            obtenerComentarios(posts[0].id, function(comentarios) {
                callback({ usuario, posts, comentarios });
            });
        });
    });
}
// Pregunta al chat: "Convierte esto a async/await"


// Ejemplo 3: Función sin validación
function crearUsuario(nombre, email, edad) {
    return {
        nombre: nombre,
        email: email,
        edad: edad,
        createdAt: new Date()
    };
}
// Pregunta al chat: "Agrega validación de parámetros y manejo de errores"


// Ejemplo 4: Código repetitivo
function calcularDescuentoBasico(precio) {
    return precio * 0.9;
}

function calcularDescuentoPremium(precio) {
    return precio * 0.8;
}

function calcularDescuentoVIP(precio) {
    return precio * 0.7;
}
// Pregunta al chat: "Refactoriza usando un objeto de configuración"


// Ejemplo 5: Función sin documentación JSDoc
function filtrarProductos(productos, categoría, precioMin, precioMax) {
    return productos.filter(p => 
        p.categoria === categoría && 
        p.precio >= precioMin && 
        p.precio <= precioMax
    );
}
// Pregunta al chat: "Agrega documentación JSDoc completa"


// Ejemplo 6: Clase que necesita mejoras
class CarritoCompras {
    constructor() {
        this.items = [];
    }
    
    agregar(item) {
        this.items.push(item);
    }
    
    calcularTotal() {
        let total = 0;
        for (let item of this.items) {
            total += item.precio;
        }
        return total;
    }
}
// Pregunta al chat: "Agrega métodos para remover items, aplicar descuentos y gestionar cantidad"


// Ejemplo 7: Función que necesita optimización
function buscarEnArray(arr, valor) {
    for (let i = 0; i < arr.length; i++) {
        if (arr[i] === valor) {
            return i;
        }
    }
    return -1;
}
// Pregunta al chat: "¿Hay una forma más eficiente de hacer esto con métodos nativos?"


// Ejemplo 8: Manejo de estado sin validación
let appState = {
    user: null,
    isAuthenticated: false
};

function updateState(newState) {
    appState = { ...appState, ...newState };
}
// Pregunta al chat: "Convierte esto en un gestor de estado más robusto con validación"


// Ejemplo 9: Función asíncrona sin manejo de errores
async function fetchUserData(userId) {
    const response = await fetch(`/api/users/${userId}`);
    const data = await response.json();
    return data;
}
// Pregunta al chat: "Agrega manejo de errores completo y reintentos"


// Ejemplo 10: Código imperativo que podría ser funcional
function procesarPedidos(pedidos) {
    let resultado = [];
    for (let pedido of pedidos) {
        if (pedido.estado === 'pendiente') {
            let pedidoActualizado = {
                ...pedido,
                estado: 'procesando',
                fechaProceso: new Date()
            };
            resultado.push(pedidoActualizado);
        }
    }
    return resultado;
}
// Pregunta al chat: "Refactoriza usando programación funcional (map, filter, reduce)"


// Ejemplo 11: Component React básico
function UserProfile({ user }) {
    return (
        <div>
            <h1>{user.name}</h1>
            <p>{user.email}</p>
            <p>{user.bio}</p>
        </div>
    );
}
// Pregunta al chat: "Agrega PropTypes, manejo de loading y error states"


// Ejemplo 12: Validación de formulario básica
function validarFormulario(datos) {
    if (!datos.email) return false;
    if (!datos.password) return false;
    if (datos.password.length < 6) return false;
    return true;
}
// Pregunta al chat: "Mejora esta validación con mensajes de error específicos y regex para email"


// Ejemplo 13: Utilidad que necesita tests
function formatearFecha(fecha, formato) {
    const opciones = {
        'corta': { day: '2-digit', month: '2-digit', year: 'numeric' },
        'larga': { day: 'numeric', month: 'long', year: 'numeric' },
        'completa': { weekday: 'long', day: 'numeric', month: 'long', year: 'numeric' }
    };
    
    return new Date(fecha).toLocaleDateString('es-ES', opciones[formato]);
}
// Pregunta al chat: "Genera una suite de tests unitarios con Jest"


// Ejemplo 14: API client sin estructura
function llamarAPI(endpoint, metodo, datos) {
    return fetch(endpoint, {
        method: metodo,
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(datos)
    }).then(res => res.json());
}
// Pregunta al chat: "Crea una clase API client completa con interceptors y manejo de errores"


// Ejemplo 15: Lógica de negocio sin separación de responsabilidades
function procesarCompra(usuario, producto, cantidad) {
    // Validar stock
    if (producto.stock < cantidad) {
        alert('Stock insuficiente');
        return false;
    }
    
    // Calcular precio
    let precio = producto.precio * cantidad;
    
    // Aplicar descuento
    if (usuario.tipo === 'premium') {
        precio = precio * 0.9;
    }
    
    // Actualizar stock
    producto.stock -= cantidad;
    
    // Guardar en BD
    fetch('/api/compras', {
        method: 'POST',
        body: JSON.stringify({ usuario, producto, cantidad, precio })
    });
    
    alert('Compra exitosa');
    return true;
}
// Pregunta al chat: "Refactoriza separando responsabilidades y usando arquitectura limpia"

File: ejemplos-chat.py
"""
Ejemplos de código para practicar con GitHub Copilot Chat
Usa estos ejemplos para hacer preguntas al chat y mejorar tu código
"""

# Ejemplo 1: Función básica que necesita mejoras
def procesar_datos(datos):
    resultado = []
    for d in datos:
        if d > 0:
            resultado.append(d * 2)
    return resultado

# Pregunta al chat: "¿Cómo puedo mejorar esta función usando comprensión de listas?"


# Ejemplo 2: Función con posibles errores
def dividir_numeros(a, b):
    return a / b

# Pregunta al chat: "¿Qué errores podría tener esta función?"


# Ejemplo 3: Código que necesita documentación
def calcular_estadisticas(numeros):
    total = sum(numeros)
    promedio = total / len(numeros)
    maximo = max(numeros)
    minimo = min(numeros)
    return {
        'total': total,
        'promedio': promedio,
        'maximo': maximo,
        'minimo': minimo
    }

# Pregunta al chat: "Agrega documentación completa a esta función"


# Ejemplo 4: Algoritmo ineficiente
def buscar_duplicados(lista):
    duplicados = []
    for i in range(len(lista)):
        for j in range(i + 1, len(lista)):
            if lista[i] == lista[j] and lista[i] not in duplicados:
                duplicados.append(lista[i])
    return duplicados

# Pregunta al chat: "Optimiza esta función para O(n) usando un set"


# Ejemplo 5: Función que necesita manejo de errores
def leer_archivo_json(ruta):
    import json
    with open(ruta, 'r') as archivo:
        datos = json.load(archivo)
    return datos

# Pregunta al chat: "Agrega manejo de errores completo a esta función"


# Ejemplo 6: Clase que necesita métodos adicionales
class Usuario:
    def __init__(self, nombre, email):
        self.nombre = nombre
        self.email = email

# Pregunta al chat: "Agrega métodos para validar email y representación en string"


# Ejemplo 7: Código que necesita tests
def validar_password(password):
    if len(password) < 8:
        return False
    tiene_numero = any(c.isdigit() for c in password)
    tiene_mayuscula = any(c.isupper() for c in password)
    tiene_minuscula = any(c.islower() for c in password)
    return tiene_numero and tiene_mayuscula and tiene_minuscula

# Pregunta al chat: "Genera tests unitarios completos para esta función"


# Ejemplo 8: Función que puede usar tipado
def combinar_listas(lista1, lista2):
    resultado = []
    for item1 in lista1:
        for item2 in lista2:
            resultado.append((item1, item2))
    return resultado

# Pregunta al chat: "Agrega type hints a esta función"


# Ejemplo 9: Código repetitivo que necesita refactorización
def procesar_usuario_admin(usuario):
    if usuario['rol'] == 'admin':
        print(f"Procesando admin: {usuario['nombre']}")
        usuario['permisos'] = ['leer', 'escribir', 'eliminar']
        usuario['nivel'] = 10
        return usuario

def procesar_usuario_editor(usuario):
    if usuario['rol'] == 'editor':
        print(f"Procesando editor: {usuario['nombre']}")
        usuario['permisos'] = ['leer', 'escribir']
        usuario['nivel'] = 5
        return usuario

def procesar_usuario_lector(usuario):
    if usuario['rol'] == 'lector':
        print(f"Procesando lector: {usuario['nombre']}")
        usuario['permisos'] = ['leer']
        usuario['nivel'] = 1
        return usuario

# Pregunta al chat: "Refactoriza estas funciones usando un patrón más eficiente"


# Ejemplo 10: Código que necesita ser asíncrono
def obtener_datos_api(urls):
    import requests
    resultados = []
    for url in urls:
        respuesta = requests.get(url)
        resultados.append(respuesta.json())
    return resultados

# Pregunta al chat: "Convierte esta función a async/await para mejor rendimiento"
